src/Components/Web/src/Forms/InputBase.cs

```diff
@@ -216,8 +217,30 @@ public override Task SetParametersAsync(ParameterView parameters)
                     $"{nameof(Forms.EditContext)} dynamically.");
             }
 
+            SetAdditionalAttributesIfValidationFailed();
+
             // For derived components, retain the usual lifecycle with OnInit/OnParametersSet/etc.
-            return base.SetParametersAsync(ParameterView.Empty);
+            return base.SetParametersAsync(parameters);
+        }
+
+        private void OnValidateStateChanged(object sender, ValidationStateChangedEventArgs eventArgs)
+        {
+            SetAdditionalAttributesIfValidationFailed();
+
+            StateHasChanged();
+        }
+
+        private void SetAdditionalAttributesIfValidationFailed()
+        {
+            if (!EditContext.GetValidationMessages(FieldIdentifier).Any())
+                return;
+
+            if (AdditionalAttributes == null)
+                AdditionalAttributes = new Dictionary<string, object>();
+
+            // To make the `Input` components accessible by default
+            // we will automatically render the `aria-invalid` attribute when the validation fails
+            AdditionalAttributes["aria-invalid"] = true;
```

Comment:
As per comments elsewhere, it could be good to check if `AdditionalAttributes` already contains an entry for `aria-invalid`, and if so don't overwrite it.

Also, minor code-style issues in this method, just for consistency with the rest of the repo:

 * Please use braces even for single-line `if` statements
 * This one's more debatable, but it's relatively unusual that we'd drop a `return` statement in where it appears on L236, and more common for us to invert that condition so the whole method body is inside an `if (invalid) { ... }` without the need for a `return`.
    * This is debatable and I won't argue if you want to keep it as-is. The potential advantage of avoiding the early `return` is that someone reading the code can see just from the indentation that it's all contingent on the `if` condition, and you don't have to read backwards to reason about why it's OK for L243 to assume the state is invalid.
    * If you do prefer to leave it as-is, I'd recommend adding a comment just before line 238 saying something like `// By now, we know the field state is invalid`, so it can be skim-read quicker.

Reply:
Does it need to be the same instance? Can we avoid changing the API and set the property instead?

```
var additionalAttributes = AdditionalAttributes?.ToDictionary() ?? new Dictionary<string, object>(...);

AdditionalAttributes = additionalDictionary;
```

Avoids a breaking change in the API

Reply:
Unfortunately that would be more allocatey.

I guess we could check if the underlying object was actually cast able to Dictionary and if so do that (assuming it actually is in standard cases given that the framework supplies it), otherwise fall back on allocating and copying.

Reply:
> I suppose the most common case will be it being null and we’d be allocating then anyway.

👍 

It seems iffy to make an API breaking change to save a possibly dictionary allocation. Downcasting should avoid it regardless. In the most common case, it's a regular dictionary instance.